src/main/java/com/google/devtools/build/lib/unix/UnixFileSystem.java

```diff
@@ -409,6 +414,14 @@ public void setLastModifiedTime(Path path, long newTime) throws IOException {
 
   @Override
   protected byte[] getDigest(Path path) throws IOException {
+    // First attempt to obtain the digest from an extended attribute attached to the file. This
```

Comment:
I think this should probably be in getFastDigest, not getDigest. This should definitely be consistent with how it works Google-internal.

Reply:
Well, that's something I'm wondering about whether that's desirable. If I understand it correctly, FileSystem.getDigest() and FileSystem.getFastDigest() are eventually called through [DigestUtils.getDigestOrFail()](https://github.com/bazelbuild/bazel/blob/afc31f3391acc7f38b47f9229309a97419d88f8b/src/main/java/com/google/devtools/build/lib/actions/cache/DigestUtils.java#L224-L258). That function roughly does this:

1. Call FileSystem.getFastDigest().
2. If that fails, do an in-memory cache lookup.
3. If that fails, call FileSystem.getDigest() and insert the results into the in-memory cache.

If we were to place this logic in FileSystem.getFastDigest(), doesn't that mean we'll see a large increase in system calls due to us no longer caching digests in memory?

Reply:
That cache is not used by Google internally because of its correctness issues, it's only used by Bazel. If digests have a small constant-time overhead, then I would recommend not using that cache anyway. I think it only comes into play when Bazel has to discard its (correct) in-memory state because of an analysis flag change and is then doing an action-cached build. So I agree with @ulfjack here.

Reply:
Caching of source file digests happens in Skyframe as part of `ArtifactFunction` even if the analysis phase is dropped, if I'm not mistaken.

Reply:
That's correct, but I think this cache was added in https://github.com/bazelbuild/bazel/commit/7ac6d59bcd8ea489ddb4218b495e131346419302 by @jmmv for output files, which are cleared on analysis changes.

Reply:
Thanks for clarifying! I've just changed the patch to move the getxattr() call to getFastDigest().